src/app/data-model-provider/tests/TestActionReturnStatus.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/core/CHIPError.h"
+#include "protocols/interaction_model/StatusCode.h"
+#include "pw_unit_test/framework_backend.h"
+#include <app/data-model-provider/ActionReturnStatus.h>
+#include <app/data-model-provider/StringBuilderAdapters.h>
+#include <lib/support/CodeUtils.h>
+
+#include <pw_unit_test/framework.h>
+
+using chip::app::DataModel::ActionReturnStatus;
+using chip::Protocols::InteractionModel::ClusterStatusCode;
+using chip::Protocols::InteractionModel::Status;
+
+TEST(TestActionReturnStatus, TestEquality)
+{
+    // equality should happen between equivalent statuses and chip_errors
+    ASSERT_EQ(ActionReturnStatus(Status::UnsupportedRead), Status::UnsupportedRead);
+    ASSERT_EQ(ActionReturnStatus(Status::UnsupportedWrite), CHIP_IM_GLOBAL_STATUS(UnsupportedWrite));
+
+    ASSERT_EQ(ActionReturnStatus(CHIP_IM_GLOBAL_STATUS(Busy)), Status::Busy);
+    ASSERT_EQ(ActionReturnStatus(CHIP_IM_GLOBAL_STATUS(Busy)), CHIP_IM_GLOBAL_STATUS(Busy));
+
+    ASSERT_EQ(ActionReturnStatus(CHIP_IM_CLUSTER_STATUS(123)), CHIP_IM_CLUSTER_STATUS(123));
+    ASSERT_EQ(ActionReturnStatus(ClusterStatusCode::ClusterSpecificFailure(123)), CHIP_IM_CLUSTER_STATUS(123));
+    ASSERT_EQ(ActionReturnStatus(ClusterStatusCode::ClusterSpecificFailure(123)), ClusterStatusCode::ClusterSpecificFailure(123));
+    ASSERT_EQ(ActionReturnStatus(ClusterStatusCode::ClusterSpecificSuccess(123)), ClusterStatusCode::ClusterSpecificSuccess(123));
+
+    // Successes (without cluster-specific codes) are equivalent
+    ASSERT_EQ(ActionReturnStatus(Status::Success), Status::Success);
+    ASSERT_EQ(ActionReturnStatus(CHIP_NO_ERROR), Status::Success);
+    ASSERT_EQ(ActionReturnStatus(Status::Success), CHIP_NO_ERROR);
+
+    // status specific success is has more data, so there is no equality (i.e. an action return
+    // with specific success codes has more data than a simple success)
+    ASSERT_NE(ActionReturnStatus(ClusterStatusCode::ClusterSpecificSuccess(123)), CHIP_NO_ERROR);
+    ASSERT_NE(ActionReturnStatus(ClusterStatusCode::ClusterSpecificSuccess(123)), Status::Success);
+    ASSERT_NE(ActionReturnStatus(CHIP_NO_ERROR), ClusterStatusCode::ClusterSpecificSuccess(123));
+    ASSERT_NE(ActionReturnStatus(Status::Success), ClusterStatusCode::ClusterSpecificSuccess(123));
```

Comment:
Could you add a case to compare two cluster-specific successes with different success codes?

Reply:
Added a lot more tests ... still find it somewhat hard to pick all combinations. Hopefully I have solid coverage.